compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/IncrementalCompilerRunner.kt

```diff
@@ -75,192 +86,210 @@ abstract class IncrementalCompilerRunner<
+            } catch (e: Throwable) {
+                return ICResult.Failed(IC_FAILED_TO_GET_CHANGED_FILES, e)
             }
-            val allKotlinFiles = allSourceFiles.filter { it.isKotlinFile(kotlinSourceFilesExtensions) }
-            return compileIncrementally(
-                args, caches, allKotlinFiles, CompilationMode.Rebuild(reason), messageCollector, withAbiSnapshot,
-                classpathAbiSnapshot = classpathAbiSnapshot
-            ).also {
-                if (it == ExitCode.OK) {
-                    performWorkAfterSuccessfulCompilation(caches, wasIncremental = false)
+
+            val classpathAbiSnapshot = if (withAbiSnapshot) getClasspathAbiSnapshot(args) else null
+
+            // Step 2: Compute files to recompile
+            val compilationMode = try {
+                reporter.measure(BuildTime.IC_CALCULATE_INITIAL_DIRTY_SET) {
+                    calculateSourcesToCompile(caches, changedFiles, args, messageCollector, classpathAbiSnapshot ?: emptyMap())
                 }
+            } catch (e: Throwable) {
+                return ICResult.Failed(IC_FAILED_TO_COMPUTE_FILES_TO_RECOMPILE, e)
+            }
+
+            if (compilationMode is CompilationMode.Rebuild) {
+                return ICResult.RequiresRebuild(compilationMode.reason)
+            }
+
+            if (withAbiSnapshot && !abiSnapshotFile.exists()) {
+                reporter.debug { "Jar snapshot file does not exist: ${abiSnapshotFile.path}" }
+                return ICResult.RequiresRebuild(NO_ABI_SNAPSHOT)
+            }
+            if (withAbiSnapshot) {
+                reporter.info { "Incremental compilation with ABI snapshot enabled" }
+            }
+            val abiSnapshotData = if (withAbiSnapshot) {
+                AbiSnapshotData(
+                    snapshot = AbiSnapshotImpl.read(abiSnapshotFile),
+                    classpathAbiSnapshot = classpathAbiSnapshot!!
+                )
+            } else null
+
+            // Step 3: Compile incrementally
+            exitCode = try {
+                compileImpl(compilationMode as CompilationMode.Incremental, allSourceFiles, args, caches, abiSnapshotData, messageCollector)
+            } catch (e: Throwable) {
+                return ICResult.Failed(IC_FAILED_TO_COMPILE_INCREMENTALLY, e)
+            }
+        } catch (e: Throwable) {
+            // Because `caches` is a Closeable resource, it is good practice to close them in the event of an exception (in addition to
+            // closing them after a normal use).
+            try {
+                caches.close()
+            } catch (e2: Throwable) {
+                e.addSuppressed(e2)
             }
-        } finally {
-            caches.close(true)
+            throw e
+        }
+        try {
+            caches.close()
+        } catch (e: Throwable) {
+            return ICResult.Failed(IC_FAILED_TO_CLOSE_CACHES, e)
         }
```

Comment:
Right, this is needed only because closing caches still has bugs and throws exceptions. Once we've fixed those bugs, all of this will go away so we can just write:
```
createCacheManager(args, projectDir).use { caches ->
   ...
}
```

just like the code in `compileNonIncrementally`.